[FIX] web: retain 'All Calendar' filter on breadcrumb jump

sura-odoo opened the pull request:
Version:
----------
sass-16.3

Steps to reproduce:
--------------------------
- open the Calendar module
- click on the everybody's calendar filter option to view all the meetings
- click on any meeting to open the form view for that meeting
- now navigate back through the breadcrumb
- the everybody's calendar filter is now unchecked

Issue:
----------
The 'Everybody's Calendar' option is not set and one needs to check it again every time he come back through the breadcrumb.

Cause:
----------
​All the other filters are saved in the backend through orm calls but the All filter is not saved since it does not have a recordId.

Solution:
----------
The filter state is now stored in the user context, so when a user come back through the breadcrumb the previous set value is taken from the user context.


task-3792066

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

kpt-odoo commented:
Thanks for the fix! 😊 Just a small remark:
For this fix, we need a test to ensure that it works correctly.

Also, please replace 'calendar' in the PR/commit title messages with 'web module' since your changes are in the web module.

Additionally, PR/commit messages should start with a verb, which is correct in your case, but avoid using continuing verbs like 'losing' if I'm not mistaken.

The message should specify what the PR aims for, not just what the issue was.

You can refer to this link for more info: [Odoo Git Guidelines](https://www.odoo.com/documentation/17.0/contributing/development/git_guidelines.html) 📝

sura-odoo commented:
Thanks for highlighting the improvements.
I am done with the updates from my side.

aab-odoo reviewed (commented):
@mcm-odoo Could you review this?

I seriously doubt the user context is the place to store that information...

sura-odoo commented:
I have implemented the necessary modifications. The "All Filter" state is now added to the local view state, rather than the user context. Let me know if anything needs to be changed. Thanks :slightly_smiling_face:

mcm-odoo commented on addons/web/static/src/views/calendar/filter_panel/calendar_filter_panel.js:
```diff
@@ -19,6 +20,14 @@ export class CalendarFilterPanel extends Component {
         this.addDialog = useOwnedDialogs();
         this.orm = useService("orm");
+
+        useSetupView({
+            getLocalState: () => {
+                const sections = this.props.model.filterSections;
+                const allFilter = sections.length > 0 ? sections[0].filters?.find((f) => f.type === "all") : false;
+                return allFilter ? { allFilterState: allFilter.active } : {};
+            },
+        });
```
We are not in a view, don't use `useSetupView` here. Set the state in `CalendarModel.updateFilter` instead.

mcm-odoo commented on addons/web/static/src/views/calendar/calendar_model.js:
```diff
@@ -818,7 +818,7 @@ export class CalendarModel extends Model {
                 ? this.env._t("Everybody's calendars")
                 : this.env._t("Everything"),
-            active: previousAllFilter ? previousAllFilter.active : false,
+            active: previousAllFilter ? previousAllFilter.active : this.meta.allFilterLocalState,
```
We can have x number of sections so using a simple boolean will set the filter in every section.

mcm-odoo commented on addons/web/static/tests/views/calendar/calendar_view_tests.js:
```diff
@@ -4921,4 +4921,52 @@ QUnit.module("Views", ({ beforeEach }) => {
+        await doAction(webClient, 1);
+
+        await click(target, ".o_calendar_filter_item[data-value='all'] input");
+        await click(target, ".o_cp_switch_buttons .o_list");
```
Assert that the input is checked before switching the view.

sura-odoo commented:
Hello @mcm-odoo ,
Now I am setting the filter state in calendar model and as for the problem that we can have x number of  sections now I am uniquely identifying the filters on their section's label

mcm-odoo reviewed (commented):
Hello, I propose you to make `allFilter` a `dict` where the keys are the section labels.
With `allFilter = { sectionLabel, allFilterValue }` you know which filter is changed but you only save one filter. In the case we have multiple section with "all filter", if you check them all then only the last one will be saved when leaving the view. All the other will be reset to the default state when coming back in the view.

sura-odoo commented:
Hello @mcm-odoo,
Thanks for the suggestions
I have applied the necessary changes :smiley:
